### PR TITLE
[risk=no][RW-7563] Add missing cohort access tests

### PR DESCRIPTION
**Pull request by chenchals**
Description:
Added test cases for CohortsController for different WorkspaceAccessLevel cases
- createCohort
- duplicateCohort
- updateCohort
- deleteCohort
---
**PR checklist**

- [ ] This PR meets the Acceptance Criteria in the JIRA story
- [ ] The JIRA story has been moved to Dev Review
- [ ] This PR includes appropriate unit tests
- [ ] I have run and tested this change locally
- [ ] I have run the E2E tests on ths change against my local UI and/or API server with `yarn test-local` or [yarn test-local-devup](https://github.com/all-of-us/workbench/blob/main/e2e/README.md#examples)
- [ ] If this includes a UI change, I have taken screen recordings or screenshots of the new behavior and notified the PO and UX designer
- [ ] If this includes an API change, I have updated the appropriate Swagger definitions and notified API consumers
- [ ] If this includes a new feature flag, I have created and linked new JIRA tickets to (a) turn on the feature flag and (b) remove it later

### CODE REVIEW

File: api/src/test/java/org/pmiops/workbench/api/CohortsControllerTest.java [hunk 1]
```diff
@@ -426,6 +432,39 @@ public void testGetCohortsInWorkspace() throws Exception {
     assertThat(cohorts.size()).isEqualTo(2);
   }
 
+  @Test
+  public void testCreateCohortAccessLevelOwner() {
+    Cohort saved = createTestCohort(WorkspaceAccessLevel.OWNER);
```
Comment (freemabd) [inline api/src/test/java/org/pmiops/workbench/api/CohortsControllerTest.java:437]: It's not clear that your testing the controller#createCohort method with the logic buried in createTestCohort method. I have to drill into this method to figure out what is going on. By the name of the method I'm assuming it just creates a test Cohort object, but in reality it is doing much more(stubbing workspace and workspace acl and calling a controller method). I prefer giving this a more descriptive name or splitting it into more than one method, something like stubWorkspaceAndAcl(), createTestCohort() and callControllerCreateCohort()

File: api/src/test/java/org/pmiops/workbench/api/CohortsControllerTest.java [hunk 2]
```diff
@@ -426,6 +432,39 @@ public void testGetCohortsInWorkspace() throws Exception {
     assertThat(cohorts.size()).isEqualTo(2);
   }
 
+  @Test
+  public void testCreateCohortAccessLevelOwner() {
```
Comment (freemabd) [inline api/src/test/java/org/pmiops/workbench/api/CohortsControllerTest.java:436]: nit: prefer to start test names with Controller method name Ex: createCohortOwner

File: api/src/test/java/org/pmiops/workbench/api/CohortsControllerTest.java [hunk 3]
```diff
@@ -440,61 +479,185 @@ public void testCreateCohortBadCriteria() {
...
-  public void testUpdateCohort() throws Exception {
-    Cohort cohort = createDefaultCohort();
-    cohort =
+  public void testUpdateCohortAccessLevelOwner() {
+    Cohort saved = createTestCohort(WorkspaceAccessLevel.WRITER);
+    // update
+    Cohort toUpdate = saved.name(UPDATED_COHORT_NAME);
+    CLOCK.increment(1000); // lets say time ticked 1 sec past
+    cohortsController
+        .updateCohort(workspace.getNamespace(), workspace.getId(), toUpdate.getId(), toUpdate)
```
Comment (freemabd) [inline api/src/test/java/org/pmiops/workbench/api/CohortsControllerTest.java:488]:
```suggestion
    Cohort updated = cohortsController
        .updateCohort(workspace.getNamespace(), workspace.getId(), toUpdate.getId(), toUpdate)
```

File: api/src/test/java/org/pmiops/workbench/api/CohortsControllerTest.java [hunk 4]
```diff
@@ -440,61 +479,185 @@ public void testCreateCohortBadCriteria() {
...
+        .getBody();
+    Cohort updated =
         cohortsController
-            .createCohort(workspace.getNamespace(), workspace.getId(), cohort)
+            .getCohort(workspace.getNamespace(), workspace.getId(), saved.getId())
             .getBody();
```
Comment (freemabd) [inline api/src/test/java/org/pmiops/workbench/api/CohortsControllerTest.java:493]: Unnecessary call. cohortsController#updateCohort will return the updated cohort.

File: api/src/test/java/org/pmiops/workbench/api/CohortsControllerTest.java [hunk 5]
```diff
@@ -440,61 +479,185 @@ public void testCreateCohortBadCriteria() {
...
+    // update
+    stubGetWorkspace(
+        workspace.getNamespace(), workspace.getName(), CREATOR_EMAIL, WorkspaceAccessLevel.WRITER);
+    stubGetWorkspaceAcl(
+        workspace.getNamespace(), workspace.getName(), CREATOR_EMAIL, WorkspaceAccessLevel.WRITER);
+    Cohort toUpdate = saved.name(UPDATED_COHORT_NAME);
+    CLOCK.increment(1000); // lets say time ticked 1 sec past
+    cohortsController
+        .updateCohort(workspace.getNamespace(), workspace.getId(), toUpdate.getId(), toUpdate)
+        .getBody();
```
Comment (freemabd) [inline api/src/test/java/org/pmiops/workbench/api/CohortsControllerTest.java:512]:
```suggestion
    Cohort updated = cohortsController
        .updateCohort(workspace.getNamespace(), workspace.getId(), toUpdate.getId(), toUpdate)
        .getBody();
```

File: api/src/test/java/org/pmiops/workbench/api/CohortsControllerTest.java [hunk 6]
```diff
@@ -440,61 +479,185 @@ public void testCreateCohortBadCriteria() {
...
     Cohort updated =
         cohortsController
-            .updateCohort(workspace.getNamespace(), workspace.getId(), cohort.getId(), cohort)
+            .getCohort(workspace.getNamespace(), workspace.getId(), saved.getId())
             .getBody();
```
Comment (freemabd) [inline api/src/test/java/org/pmiops/workbench/api/CohortsControllerTest.java:516]: Unnecessary call. cohortsController#updateCohort will return the updated cohort.

File: api/src/test/java/org/pmiops/workbench/api/CohortsControllerTest.java [hunk 7]
```diff
@@ -440,61 +479,185 @@ public void testCreateCohortBadCriteria() {
...
 
-    assertThat(newCohort.getName()).isEqualTo(params.getNewName());
-    assertThat(newCohort.getCriteria()).isEqualTo(originalCohort.getCriteria());
-    assertThat(newCohort.getType()).isEqualTo(originalCohort.getType());
-    assertThat(newCohort.getDescription()).isEqualTo(originalCohort.getDescription());
+  @Test
+  public void testDuplicateCohortAccessLevelOwner() {
+    Cohort original = createTestCohort(WorkspaceAccessLevel.OWNER);
+    CLOCK.increment(1000L); // increment clock before duplicating
+    Cohort duplicated = createDuplicateTestCohort(WorkspaceAccessLevel.OWNER, original);
```
Comment (freemabd) [inline api/src/test/java/org/pmiops/workbench/api/CohortsControllerTest.java:631]: This method suffers the same problem as createTestCohort(). Please split into more logical methods or be descriptive in name about what this method really does.

File: api/src/test/java/org/pmiops/workbench/api/CohortsControllerTest.java [hunk 8]
```diff
@@ -893,4 +1056,51 @@ public void testMaterializeCohortWithEverything() {
...
+        workspace.getNamespace(), workspace.getName(), CREATOR_EMAIL, workspaceAccessLevel);
+
+    DuplicateCohortRequest params = new DuplicateCohortRequest();
+    params.setNewName(DUPLICATED_COHORT_NAME);
+    params.setOriginalCohortId(originalCohort.getId());
+
+    Cohort duplicated =
+        cohortsController
+            .duplicateCohort(workspace.getNamespace(), workspace.getId(), params)
+            .getBody();
```
Comment (freemabd) [inline api/src/test/java/org/pmiops/workbench/api/CohortsControllerTest.java:1099]:
```suggestion
    return cohortsController
            .duplicateCohort(workspace.getNamespace(), workspace.getId(), params)
            .getBody();
```

**Review by freemabd — COMMENTED** [review]
Thanks for increasing test coverage. Let a few comments, mostly about making method clean and concise.

File: api/src/test/java/org/pmiops/workbench/api/CohortsControllerTest.java (hunk 2 above)
Comment (chenchals) [reply api/src/test/java/org/pmiops/workbench/api/CohortsControllerTest.java:436]: I like prefixing all test-cases with test - helps me with overall grep'ing - howerevr this is not the case in many tests, but in this class - aa tests began with test prefix so retained the prefix.

File: api/src/test/java/org/pmiops/workbench/api/CohortsControllerTest.java (hunk 1 above)
Comment (chenchals) [reply api/src/test/java/org/pmiops/workbench/api/CohortsControllerTest.java:437]: updated code to address this issue

File: api/src/test/java/org/pmiops/workbench/api/CohortsControllerTest.java (hunk 3 above)
Comment (chenchals) [reply api/src/test/java/org/pmiops/workbench/api/CohortsControllerTest.java:488]: yes we do not need to be calling getCohort. updated code

File: api/src/test/java/org/pmiops/workbench/api/CohortsControllerTest.java (hunk 5 above)
Comment (chenchals) [reply api/src/test/java/org/pmiops/workbench/api/CohortsControllerTest.java:512]: yes updated code to reflect the same

File: api/src/test/java/org/pmiops/workbench/api/CohortsControllerTest.java (hunk 6 above)
Comment (chenchals) [reply api/src/test/java/org/pmiops/workbench/api/CohortsControllerTest.java:516]: updated code and removed this call

File: api/src/test/java/org/pmiops/workbench/api/CohortsControllerTest.java (hunk 7 above)
Comment (chenchals) [reply api/src/test/java/org/pmiops/workbench/api/CohortsControllerTest.java:631]: updated code to address these issues

File: api/src/test/java/org/pmiops/workbench/api/CohortsControllerTest.java (hunk 8 above)
Comment (chenchals) [reply api/src/test/java/org/pmiops/workbench/api/CohortsControllerTest.java:1099]: removed this method and updated code...

File: api/src/test/java/org/pmiops/workbench/api/CohortsControllerTest.java [hunk 9]
```diff
@@ -426,6 +434,56 @@ public void testGetCohortsInWorkspace() throws Exception {
...
+  public void testCreateCohortOwner() {
+    stubWorkspaceAccessLevelWithCreatorEmail(workspace, WorkspaceAccessLevel.OWNER);
+    Cohort saved =
+        cohortsController
+            .createCohort(workspace.getNamespace(), workspace.getId(), createDefaultCohort())
+            .getBody();
+    assertThat(saved.getCriteria()).isEqualTo(createDefaultCohort().getCriteria());
+    assertThat(saved.getCreator()).isEqualTo(CREATOR_EMAIL);
+    assertThat(saved.getCreationTime()).isNotNull();
+    assertThat(saved.getCreationTime()).isEqualTo(saved.getLastModifiedTime());
```
Comment (freemabd) [inline api/src/test/java/org/pmiops/workbench/api/CohortsControllerTest.java:447]: Thank you, this is so much more readable.

**Review by freemabd — APPROVED** [review]
Looks good.

File: api/src/test/java/org/pmiops/workbench/api/CohortsControllerTest.java (hunk 9 above)
Comment (chenchals) [reply api/src/test/java/org/pmiops/workbench/api/CohortsControllerTest.java:447]: :-)

**Pull request merged**